File: src/main/java/me/austin/queer/modules/setting/settings/NumberSetting.java
package me.austin.queer.modules.setting.settings;

import me.austin.queer.modules.setting.Setting;

public abstract class NumberSetting<T extends Number> extends Setting<T> {
    protected T min, max;

    public NumberSetting(String name, String description, T defaultValue, T min, T max) {
        super(name, description, defaultValue);
        this.min = min;
        this.max = max;
    }

    public Number getMin() {
        return this.min;
    }

    public Number getMax() {
        return this.max;
    }

    @Override
    public void set(Number value) {
        this.set(value);
    } 
}

File: src/main/kotlin/me/austin/client/impl/setting/NumberSetting.kt
package me.austin.client.impl.setting

import me.austin.client.api.setting.Setting
import me.austin.client.util.round

abstract class NumberSetting<T : Number>(
    name: String, description: String, default: T, protected val increment: T, vararg children: Setting<*>
) : Setting<T>(name, description, default, *children)

class LongSetting(name: String, description: String, default: Long, vararg children: Setting<*>) :
    NumberSetting<Long>(name, description, default, 1L, *children)

class IntSetting(name: String, description: String, default: Int, vararg children: Setting<*>) :
    NumberSetting<Int>(name, description, default, 1, *children)

class ShortSetting(name: String, description: String, default: Short, vararg children: Setting<*>) :
    NumberSetting<Short>(name, description, default, 1, *children)

class DoubleSetting(
    name: String, description: String, default: Double, increment: Double, vararg children: Setting<*>
) : NumberSetting<Double>(name, description, default, increment, *children) {
    fun set(other: Double, round: Boolean) = if (round) this.value = other.round(this.increment) else this.value = other
}

class FloatSetting(name: String, description: String, default: Float, increment: Float, vararg children: Setting<*>) :
    NumberSetting<Float>(name, description, default, increment, *children) {
    fun set(other: Float, round: Boolean) = if (round) this.value = other.round(this.increment) else this.value = other
}